compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/interop/JMethodResolver.java

```diff
@@ -271,10 +272,33 @@ private void validateExceptionTypes(JMethodRequest jMethodRequest, JMethod jMeth
 
         if ((throwsCheckedException && !jMethodRequest.returnsBErrorType) ||
                 (jMethodRequest.returnsBErrorType && !throwsCheckedException && !returnsErrorValue)) {
+            BType returnType = jMethodRequest.bReturnType;
+            BType modifiedRetType = returnType;
+            if (returnType.tag == TypeTags.UNION) {
+                BUnionType bUnionReturnType = (BUnionType) returnType;
+                modifiedRetType = BUnionType.create(null, getNonErrorMembers(bUnionReturnType));
+            }
+            String expectedRetTypeName = modifiedRetType + "|error";
```

Comment:
We do not need to run this code if it is null. 

ex. 
if (returnType.tag == TypeTags.NIL || returnType instanceof BTypeReferenceType &&
                    ((BTypeReferenceType) returnType).referredType.tag == TypeTags.ERROR) {
                expectedRetTypeName = "error";
            } else ((returnType.tag == TypeTags.UNION)){
   BUnionType bUnionReturnType = (BUnionType) returnType;
                BType  modifiedRetType = BUnionType.create(null, getNonErrorMembers(bUnionReturnType));
String expectedRetTypeName = modifiedRetType + "|error";
} else {
 expectedRetTypeName = returnType + "|error";
}

Reply:
Addressed with https://github.com/ballerina-platform/ballerina-lang/pull/33876/commits/88f9cd6bd08f78633724cfcad49797b6cab31330